core/src/main/kotlin/xtdb/util/HyperLogLog.kt

```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * Estimate the cardinality of the set represented by this HyperLogLog
+     * @param hll The ByteBuf containing HLL registers
+     * @return The estimated cardinality
+     */
+    @JvmStatic
+    fun estimate(hll: ByteBuf): Double {
+        val m = hll.capacity() / Integer.BYTES
+
+        var acc = 0.0
+        for (i in 0 until hll.capacity() step Integer.BYTES) {
+            acc += 2.0.pow(-hll.getInt(i))
+        }
+
+        val z = 1.0 / acc
+        val alpha = 0.7213 / (1.0 + 1.079 / m)
+        val e = alpha * m.toDouble().pow(2.0) * z
+
+        // Small range correction
+        if (e <= 2.5 * m) {
+            var v = 0L
+            for (i in 0 until hll.capacity() step Integer.BYTES) {
+                if (hll.getInt(i) == 0) {
+                    v++
+                }
+            }
+
+            return if (v == 0L) {
+                e
+            } else {
+                m * ln(m.toDouble() / v)
+            }
+        }
+
+        if (e > twoTo32 / 30) {
+            return -twoTo32 * ln(1.0 - e / twoTo32)
+        }
+
+        return e
+    }
+
+    /**
+     * Create a new HyperLogLog ByteBuf with the combined maximum of two HLLs
+     * @param hllA First HLL buffer
+     * @param hllB Second HLL buffer
+     * @return A new ByteBuf containing the combined HLL
+     */
+    fun combine(hllA: ByteBuf, hllB: ByteBuf): ByteBuf {
+        require(hllA.capacity() == hllB.capacity()) { "HyperLogLog buffers must have the same capacity" }
+
+        val result = Unpooled.buffer(hllA.capacity())
+        for (i in 0 until hllA.capacity() step Integer.BYTES) {
+            result.setInt(i, max(hllA.getInt(i), hllB.getInt(i)))
+        }
+        return result
+    }
+
+    /**
+     * Estimate the cardinality of the union of two sets
+     * @param hllA First HLL buffer
+     * @param hllB Second HLL buffer
+     * @return The estimated cardinality of the union
+     */
+    fun estimateUnion(hllA: ByteBuf, hllB: ByteBuf): Double {
+        return estimate(combine(hllA, hllB))
+    }
+
+    /**
+     * Estimate the cardinality of the intersection of two sets
+     * @param hllA First HLL buffer
+     * @param hllB Second HLL buffer
+     * @return The estimated cardinality of the intersection
+     */
+    fun estimateIntersection(hllA: ByteBuf, hllB: ByteBuf): Double {
+        return estimate(hllA) + estimate(hllB) - estimateUnion(hllA, hllB)
+    }
+
+    /**
+     * Estimate the cardinality of the difference between two sets (A - B)
+     * @param hllA First HLL buffer
+     * @param hllB Second HLL buffer
+     * @return The estimated cardinality of the difference
+     */
+    fun estimateDifference(hllA: ByteBuf, hllB: ByteBuf): Double {
+        return estimate(hllA) - estimateIntersection(hllA, hllB)
+    }
+
+    @JvmStatic
+    fun createHLL(bufferSize: Int = DEFAULT_BUFFER_SIZE) = Unpooled.buffer(bufferSize)
+    @JvmStatic
+    fun toHLL(bytes: ByteArray) = Unpooled.wrappedBuffer(bytes)
+}
```

Comment:
@jarohen  I am starting to like this style the most in Kotlin. I am guessing because it's the closest to Clojure. Function where the first argument is the main data structure worked on. Happy to also go with extension functions or other idioms if you prefer.

Reply:
Agreed that this is closer to Clojure, but Kotlin != Clojure :slightly_smiling_face: 

I have a slight preference for extension functions wherever you'd typically use threading in Clojure (i.e. where the first parameter is the object being acted on) because it then allows you to use (e.g.) `?.` and dot-chaining in Kotlin, rather than needing more and more nesting - same as threading in Clojure, now I think about it.